File: spring-boot/src/main/java/springboot/controller/ProfileController.java
package springboot.controller;

import lombok.RequiredArgsConstructor;
import org.springframework.web.bind.annotation.DeleteMapping;
import org.springframework.web.bind.annotation.GetMapping;
import org.springframework.web.bind.annotation.PathVariable;
import org.springframework.web.bind.annotation.PostMapping;
import org.springframework.web.bind.annotation.RestController;
import springboot.domain.dto.ProfileDto;
import springboot.service.ProfileService;

@RestController
@RequiredArgsConstructor
public class ProfileController {
    private final ProfileService profileService;

    @GetMapping("/profiles/{username}")
    public ProfileDto getUserProfile(@PathVariable String username) {
        return this.profileService.getProfile(username);
    }

    @PostMapping("/profiles/{username}/follow")
    public ProfileDto followUser(@PathVariable String username) {
        return this.profileService.followUser(username);
    }

    @DeleteMapping("/profiles/{username}/follow")
    public ProfileDto unFollowUser(@PathVariable String username) {
        return this.profileService.unFollowUser(username);
    }
}

File: spring-boot-kt/src/main/kotlin/real/world/springbootkt/domain/profile/ProfileController.kt
package real.world.springbootkt.domain.profile

import org.springframework.web.bind.annotation.*

@RestController
@RequestMapping("/api")
class ProfileController(
    private val profileService: ProfileService
) {
    @GetMapping("/profiles/{username}")
    fun profileByUsername(@PathVariable(name = "username") username: String): ProfileResource.Response =
        ProfileResource.Response(profileService.getProfileByUsername(username))

    @PostMapping("/profiles/{username}/follow")
    fun follow(@PathVariable(name = "username") username: String): ProfileResource.Response =
        ProfileResource.Response(profileService.follow(username))

    @DeleteMapping("/profiles/{username}/follow")
    fun unfollow(@PathVariable(name = "username") username: String): ProfileResource.Response =
        ProfileResource.Response(profileService.unfollow(username))

}